core/scratch_block_comment.js

```diff
@@ -39,18 +39,35 @@ goog.require('goog.userAgent');
 /**
  * Class for a comment.
  * @param {!Blockly.Block} block The block associated with this comment.
+ * @param {string} text The text content of this comment.
+ * @param {string=} id Optional uid for comment; a new one will be generated if
+ *     not provided.
  * @param {number=} x Initial x position for comment, in workspace coordinates.
  * @param {number=} y Initial y position for comment, in workspace coordinates.
  * @param {boolean=} minimized Whether or not this comment is minimized
  *     (only the top bar displays), defaults to false.
  * @extends {Blockly.Comment}
  * @constructor
  */
-Blockly.ScratchBlockComment = function(block, x, y, minimized) {
+Blockly.ScratchBlockComment = function(block, text, id, x, y, minimized) {
   Blockly.ScratchBlockComment.superClass_.constructor.call(this, block);
+  this.text_ = text;
   this.x_ = x;
   this.y_ = y;
   this.isMinimized_ = minimized || false;
+
+  this.workspace = block.workspace;
+  this.id = goog.isString(id) && !this.workspace.getCommentById(id) ?
+      id : Blockly.utils.genUid();
+  this.workspace.addTopComment(this);
```

Comment:
Nit: should be workspace.addComment, since "top comments" no longer means anything.  We can fix that later.

Reply:
#1564